dev_localSearch_lhcSampling_nb101dictCreation

Stargazer1998 opened the pull request:
Add a script for the creation of Nas101 dict. creator and helper functions to be used

## Describe your changes
Common functions to be used for LHC Sampler and nb101 dict creation
nb101 dictionary creator script 
LCH Sampler Script 
modify local search to work with sampled points and nb101 dict. 
(local search tests are not active for now )
## What is the related task number?
-
## Checklist before requesting a review
- [x] I have performed a self-review of my code
- [ ] If it is a core feature, I have added thorough tests.

Stargazer1998 committed:
add nb101 dictionary creator

Add a script for the creation of Nas101 dict. creator and helper functions to be used

emreds commented on src/tum_dlr_automl_for_eo/scripts/nb101_dict_creator.py:
```diff
@@ -0,0 +1,36 @@
+
+
+def create_encoded_arc_file():
+    nasbench = api.NASBench(path_to_read_data, data_format = data_format)
```
Why we are iterating over all the architectures in nasbench here?

Stargazer1998 replied:
We would like to create a kind of look-up table for each nb101 architecture, as {encoded_arc: nb101_arc}. Thus, we do not need to decode again during neighbour search.

emreds replied:
I see makes sense :+1:

Stargazer1998 committed:
Add Latin Hyper Cube Sampler

Add LHC Sampler script
Modify nb101_dict_creator to store accuracy of each architectures
Add another helper function to reach from key to binary encoded arc

Stargazer1998 committed:
lhc sampler / dict creator corrections

* recorrect dict to contain all nb101 keys
* sample only (N) nb101 keys
* place scripts to correct file

Stargazer1998 committed:
local search modification

* integrate local search to train new nei.
* give (LHC)sampled points as keys and read acc. from nb101 dictionary
* disable local search test for now !!

Stargazer1998 committed:
Revert "local search modification"

This reverts commit bdf89ef44f5f781543d9be744e4a577736525848.

Stargazer1998 committed:
Revert "Revert "local search modification""

This reverts commit b094b24f99e33257a471da97cbe0371641343629.

Stargazer1998 committed:
fix import

fix import in custom metric test

Stargazer1998 committed:
fix import

fix import in cutom_metrics

emreds commented on src/tum_dlr_automl_for_eo/utils/helper_functions.py:
```diff
@@ -0,0 +1,155 @@
+
+
+def sample_single_op(NUM_VERTICES, ALLOWED_OPS):
+    x = lhs(NUM_VERTICES, samples=1)[0]
```
I could not find the definition of the `lhc` function. Could you show it to me? @Stargazer1998

Stargazer1998 replied:
Upps, I should have been more clear. It is a function from pyDOE lib.

emreds replied:
Hmm then we can import it as `import pyDOE` and use it like `pyDOE.lhs()`. 
Could you add the new required packages to the requirements.txt also?

Stargazer1998 replied:
sure

Stargazer1998 committed:
add pyDOE to requirements

modify lhs function call, and add it to requirements

emreds commented on src/tum_dlr_automl_for_eo/utils/helper_functions.py:
```diff
@@ -0,0 +1,155 @@
+    trans = dict()
+    for i, ix in enumerate(pos):
+        trans[i] = ix
+    i, j = np.nonzero(adj_matrix)
```
Hmm maybe we should also `import numpy as np` flake8 gives an error about the not imported package

Stargazer1998 committed:
Update requirements.txt

add further requirements

JahodaPaul commented:
So in general I think it's good, my only question is about using LHS sampling. In Latin hypercube sampling one must first decide how many sample points to use and for each sample point remember in which row and column the sample point was taken. However, in the current implementation, we sample the LHS function one by one, which I think maybe defeats the purpose of using LHS sampling.

JahodaPaul reviewed (commented):
we shouldn't sample lhs 1 by 1 as it defeats the purpose of LHS sampling algorithm

Stargazer1998 commented:
> we shouldn't sample lhs 1 by 1 as it defeats the purpose of LHS sampling algorithm

  Thanks for comments @JahodaPaul. Actually, this function has been provided by René, it would be better to discuss with him.

emreds commented:
If you have any solutions please implement it now because it's kind of urgent @JahodaPaul  :smile:

JahodaPaul commented:
> If you have any solutions please implement it now because it's kind of urgent @JahodaPaul smile

I think after a discussion with Safa, he now knows how to implement it and if I am not wrong he's working on fixing it.
